### PR TITLE
Sample along spline

### PR DESCRIPTION
**Pull request by brisvag**
This PR implements functions to sample an image in a grid along a spline. Main changes:

1. update the spline logic to match the current logic in teamtomo/tomogeom
    - some small changes in the setup
    - added generation of smoothly varying orientations
2. create a generic grid sampler function that takes oriented points and an image, and transforms a regular grid in space based on the points and then uses it to sample from the image
3. a glue function that takes the above and allows to just pass points and an image and get a spline-based grid sampling :)

I split 2 and 3 because this way we can generate oriented points however we like.

Example code that works (no good sample data to share easily, sorry):

```py
import napari
from morphosamplers.sampler import sample_along_spline

v = napari.Viewer()
image_layer = v.open(<some_image>)[0]
points_layer = v.add_points()
```

Then I picked a few points for the spline (on the same slice in this case, but they should work in 3 dimensions):

image

Then, simply:

```py
result = sample_along_spline(points_layer.data, image_layer.data, grid_shape=(100, 100), n_samples=1000)

v2 = napari.Viewer()
v2.add_image(result)
```

Here's the middle slice of the "straightened" membrane :)

image

I have not yet hardened the code to work with 3d+n dimensions, but it should be relatively easy.

### CODE REVIEW

File: src/morphosamplers/sampler.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import numpy as np
+
+
+from .spline import Spline3D
+
+
+def sample_with_grid(positions, orientations, image, grid_shape=(10, 10), grid_spacing=1, spline_interpolation_order=3):
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing
+    grid = np.stack(np.meshgrid(width, height, 0), axis=3).reshape(-1, 3)
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:11]: how sure are you that everything is correctly oriented here (e.g. handedness is preserved)
I'm generally not a fan of not being specific with the `indexing` keyword argument of `meshgrid()` 

some comments explaining the logic here will be useful when we come back to this in the future

File: src/morphosamplers/sampler.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+from scipy.ndimage import map_coordinates
+import numpy as np
+
+
+from .spline import Spline3D
+
+
+def sample_with_grid(positions, orientations, image, grid_shape=(10, 10), grid_spacing=1, spline_interpolation_order=3):
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:8]: I assume orientations is a scipy `Rotation` - is this correct?

File: src/morphosamplers/sampler.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing
+    grid = np.stack(np.meshgrid(width, height, 0), axis=3).reshape(-1, 3)
+    rotated = []
+    for r in orientations:
+        rotated.append(r.apply(grid))
+    rotated_shifted = rotated + positions.reshape(-1, 1, 3)
+    sampled_image = map_coordinates(image, rotated_shifted.T, order=spline_interpolation_order)
+    return sampled_image.reshape(*grid_shape, -1).T
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:17]: quite a few transposes going on here - same point as above, array handling is very implicit and I find this a bit hard to follow without firing up a console and stepping through

File: src/morphosamplers/sampler.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+
+def sample_along_spline(points, image, n_samples=100, spline_order=3, **kwargs):
+    spline = Spline3D(points=points, order=spline_order)
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:21]: consider taking an already defined spline as input

File: src/morphosamplers/spline.py [hunk 5]
```diff
@@ -53,140 +49,119 @@ def validate_number_of_points(
...
+    """3D spline model with interpolation of orientations along the spline.
+
+    Basis vectors for a local coordinate system along the spline can be calculated.
+    In this coordinate system, z vectors are tangent to the spline and the xy-plane
+    changes minimally and smoothly.
+    """
+    _rotation_sampler = PrivateAttr(Slerp)
+
+    @validator('points')
+    def _is_n_plus_3d_coordinate_array(cls, v):
```
Comment (alisterburt) [inline src/morphosamplers/spline.py:132]:
```suggestion
    def _is_n_by_3_coordinate_array(cls, v):
```

how about the above if you didn't like `is_3d_coordinate_array`? n plus three doesn't make much sense to me

File: src/morphosamplers/utils.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+from typing import Optional
+
+import numpy as np
+
+
+def calculate_y_vectors_from_z_vectors(
```
Comment (alisterburt) [inline src/morphosamplers/utils.py:6]: noice

File: tests/test_spline_model.py [hunk 7]
```diff
@@ -121,6 +121,6 @@ def test_invalid_spline_derivatives(derivative_order):
         _ = spline_model.sample_spline(u=0, derivative_order=derivative_order)
 
     with pytest.raises(ValueError):
-        _ = spline_model._get_equidistance_spline_samples(
+        _ = spline_model._get_equidistant_spline_samples(
```
Comment (alisterburt) [inline tests/test_spline_model.py:124]: fyi I renamed this equidistance because I don't think equidistant makes sense for what this spline does

If this was a spline which directly gave you equidistant samples along the original spline then `equidistant_spline` would be a good name - it doesn't do that though, it gives you u values which yield equidistant values from the raw spline -> the spline confers the property of 'equidistance' on u values, hence preference for 'equidistance'

File: src/morphosamplers/sampler.py (hunk 1 above)
Comment (alisterburt) [reply src/morphosamplers/sampler.py:11]: sorry for the nit but I do think this is super important to get these right from the beginning as these samplers are designed for quantifying morphological stuff

**Review by alisterburt — COMMENTED** [review]
hey @brisvag - functionality is looking really great! few general comments and a few specific comments throughout the PR

- could you separate out the changes to the spline model from the sampling code? There are some changes there I preferred in @kevinyamauchi's version - will be easier to discuss separately if you want them in
- docstrings and variable names would be useful to help understand exactly what's expected
- in sampler.py it feels like the generation of sampling grids should be separated from the actual sampling, both are useful
- the sampling interface might be more flexible if we took in an existing spline as input for sampling rather than constructing a new one in the function - would mean you can get away from duplicating spline logic in sampling code as a bonus

File: src/morphosamplers/sampler.py (hunk 1 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:11]: Good point. Not sure at all :P This was just a first iteration to get some feedback, will add details and check things now!

File: src/morphosamplers/spline.py (hunk 5 above)
Comment (brisvag) [reply src/morphosamplers/spline.py:132]: Oops, this was aleftover from trying to do `n+3d` handling.

File: tests/test_spline_model.py (hunk 7 above)
Comment (brisvag) [reply tests/test_spline_model.py:124]: I thought this renaming was a merging mistake, that's why I didn't address it, sorry for just overriding without mentioning it :P I see your point, I guess this name would be more correct for the function below which returns equally spaced stuff. I'll roll back!

File: src/morphosamplers/sampler.py (hunk 4 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:21]: I guess it's useful to have both, right?

File: src/morphosamplers/sampler.py (hunk 2 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:8]: yup, I will add some hints and stuff to clarify

**Comment by brisvag** [comment]
I went through the PR and removed a lot of accidental/unnecessary changes, as well as added docstrings and comments... In general, I was a bit trigger happy with this PR, thanks for spending the time to go through it anyways :P

>     * in sampler.py it feels like the generation of sampling grids should be separated from the actual sampling, both are useful

Sure, can split out the functions.

**Comment by brisvag** [comment]
Ok, a bunch of updates:

1. removed a lot of changes I inadvertently did to the spline code (sorry about that mess ^^')
2. split out the sampling logic into more atomic functions
3. added docstrings and comments to hopefully explain thing better
4. benefit of point 2. is that now we can sample with 3D grids as well :P

Update of the example above:
```py
import napari
from morphosamplers.sampler import sample_along_spline, Spline3D

v = napari.Viewer()
image_layer = v.open(<some_image>)[0]
points_layer = v.add_points()
```

Some points are picked manually:

image

```py
spline = Spline3D(points=points_layer.data)
sampled = sample_image_along_spline(image_layer.data, spline=spline, n_samples=150, grid_shape=(50, 50))
```

Which makes this (central slice):

image


However, we can also just sample the points themselves with 3D grids!

```
from morphosamplers.sampler import sample_image_subvolumes

pos = points_layer.data
ori = Rotation.identity(len(pos))
sampled = sample_image_subvolumes(image_layer.data, pos, ori, grid_shape=(50, 50, 50))
```

https://user-images.githubusercontent.com/23482191/197200611-fdafff8f-7f95-4990-93c8-aacaee6aa284.mp4

File: src/morphosamplers/utils.py (hunk 6 above)
Comment (kevinyamauchi) [inline src/morphosamplers/utils.py:6]: I like this function! It would be nice to add the `Parameters` and `Returns` sections here.

File: src/morphosamplers/utils.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    smoothly with increasing n.
+    """
+    # normalise z vectors and initialise y
+    z = z.copy()
+    z /= np.linalg.norm(z, axis=1, keepdims=True)
+    y = np.empty((len(z), 3))
+
+    # normalise initial y vector so that dot product is the projection
+    if initial_y_vector is None:
+        initial_y_vector = np.array([0, 0, 1])
```
Comment (kevinyamauchi) [inline src/morphosamplers/utils.py:23]: I think it would be more clear to set the default value in the kwargs (e.g., something like, `initial_y_vector`: Union[np.ndarray, Tuple[float, float, float])  = (0, 0, 1)`)

File: src/morphosamplers/spline.py [hunk 9]
```diff
@@ -163,7 +166,7 @@ def _get_equidistance_u(self, separation: float) -> np.ndarray:
         return np.linspace(0, 1 - remainder, n_points)
 
     def _get_equidistance_spline_samples(
-        self, separation: float, derivative_order: bool = 0
+        self, separation: float, derivative_order: int = 0
```
Comment (kevinyamauchi) [inline src/morphosamplers/spline.py:169]: Nice catch, thanks!

File: src/morphosamplers/spline.py [hunk 10]
```diff
@@ -7,14 +7,17 @@
...
+from .utils import calculate_y_vectors_from_z_vectors
 
 
 class NDimensionalSpline(EventedModel):
     """Model for multidimensional splines."""
 
     points: np.ndarray
     order: conint(ge=1, le=5) = 3
-    _n_spline_samples: int = 10000
+    _n_spline_samples: int = PrivateAttr(10000)
```
Comment (kevinyamauchi) [inline src/morphosamplers/spline.py:20]: Thanks for catching this!

File: src/morphosamplers/utils.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+def calculate_y_vectors_from_z_vectors(
+    z: np.ndarray,
+    initial_y_vector: Optional[np.ndarray] = None
+) -> np.ndarray:
+    """Calculate y vectors starting from z vectors.
+
+    This function will return the set of unit vectors perpendicular to the z-vectors which are
+    maximally coaxial to their neighbours. It assumes that z vectors (n, 3) vary
+    smoothly with increasing n.
```
Comment (kevinyamauchi) [inline src/morphosamplers/utils.py:14]: I _think_ this is correct?

```suggestion
    smoothly and monotonically with increasing n.
```

File: src/morphosamplers/sampler.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    spline: Spline3D,
+    n_samples=100,
+    grid_shape=(10, 10),
+    grid_spacing=(1, 1),
+    image_interpolation_order=3,
+):
+    """Extract image planes from a volume following a spline path."""
+    u = np.linspace(0, 1, n_samples)
+    pos = spline.sample_spline(u)
+    ori = spline.sample_spline_orientations(u)
```
Comment (kevinyamauchi) [inline src/morphosamplers/sampler.py:91]: I would prefer if the variable names were complete words (i.e., `positions` and `orientations`)

**Review by kevinyamauchi — COMMENTED** [review]
Nice work, @brisvag ! Can you please add some tests for the sampler functions and the new `Spline3D` class? Also, can you please add the `Parameters` and `Returns` parts of the docstring to the `sampler.py` functions?

After that, I think this should be good to go! Thanks!

File: src/morphosamplers/utils.py (hunk 11 above)
Comment (brisvag) [reply src/morphosamplers/utils.py:14]: No, you can have a "snakey" filament and it will work just fine. What won't "work" is if you have z vector `i` and z vector `i + 1` be basically unrelated, then the results will be just as unrelated... which I guess we could even leave out :P

**Comment by brisvag** [comment]
Ok, I addressed the comments and added a bunch of docstrings and type hints :)

File: src/morphosamplers/utils.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    # normalise z vectors and initialise y
+    z = z.copy()
+    z /= np.linalg.norm(z, axis=1, keepdims=True)
+    y = np.empty((len(z), 3))
+
+    # normalise initial y vector so that dot product is the projection
+    if initial_y_vector is None:
+        initial_y_vector = np.array([0, 0, 1])
+    else:
+        initial_y_vector = initial_y_vector / np.linalg.norm(initial_y_vector)
```
Comment (alisterburt) [inline src/morphosamplers/utils.py:40]:
```suggestion
    initial_y_vector = np.asarray(initial_y_vector) / np.linalg.norm(initial_y_vector)
```

File: src/morphosamplers/utils.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    yz = np.dot(initial_y_vector, z[0])  # projection of y on first z
+    y[0] = initial_y_vector - yz * z[0]  # subtract z component of initial y
+
+    # update y vectors in order
+    for i in range(len(y) - 1):
+        yz = np.dot(y[i], z[i + 1])  # projection of y on next z
+        y[i + 1] = y[i] - yz * z[i + 1]  # subtract z component of current y
+        y[i + 1] /= np.linalg.norm(
+            y[i + 1]
+        )  # normalize each iteration to prevent precision issues
```
Comment (alisterburt) [inline src/morphosamplers/utils.py:52]:
```suggestion
        y[i + 1] /= np.linalg.norm(y[i + 1])  # prevents precision issues
```

Comment (alisterburt) [reply src/morphosamplers/utils.py:52]: just a tidiness thing

File: src/morphosamplers/sampler.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+    Returns
+    -------
+    np.ndarray
+        Coordinate of points forming the 2D grid.
+    """
+    # create indices for x and y with correct spacing
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing[0]
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing[1]
+    depth = np.linspace(-1, 1, grid_shape[2]) * grid_shape[2] * grid_spacing[2]
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:68]: this yields a grid with values in the range `[-grid_shape[i], grid_shape[i]]` if `grid_spacing[i]==1]`, i.e. length is 2 *  `grid_shape[i]` - is this intended? It seems incorrect but maybe I'm missing something

File: src/morphosamplers/sampler.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        Spacing between points in the sampling grid.
+
+    Returns
+    -------
+    np.ndarray
+        Coordinate of points forming the 2D grid.
+    """
+    # create indices for x and y with correct spacing
+    width = np.linspace(-1, 1, grid_shape[0]) * grid_shape[0] * grid_spacing[0]
+    height = np.linspace(-1, 1, grid_shape[1]) * grid_shape[1] * grid_spacing[1]
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:36]: same as below, question re total length between start and end of grid

File: src/morphosamplers/sampler.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+    # convert to coordinate form (x, y, z, 3) (at z = 0)
+    grid = np.stack(np.meshgrid(width, height, 0, indexing="ij"), axis=3)
+    # drop degenerate dim
+    return einops.rearrange(grid, "x y 1 d -> x y d")
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:40]: <3

File: src/morphosamplers/sampler.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+    return grid
+
+
+def generate_sampling_coordinates(
+    sampling_grid: np.ndarray, positions: np.ndarray, orientations: Rotation
+) -> np.ndarray:
+    """
+    Copy and transform a given sampling grid onto a set of positions and orientations.
+
+    Returns an (n, *grid_shape, 3) grid of sampling points.
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:80]:
```suggestion
    Returns a (batch, *grid_shape, 3) batch of coordinate grids for sampling.
```

File: src/morphosamplers/sampler.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+    -------
+    np.ndarray
+        Coordinate of points combining all the transformed grids.
+    """
+    rotated = []
+    grid_shape = sampling_grid.shape
+    grid_coords = sampling_grid.reshape(-1, 3)
+    # apply each orientation to the grid and store the result
+    for ori in orientations:
+        rotated.append(ori.apply(grid_coords))
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:101]:
```suggestion
    for orientation in orientations:
        rotated.append(orientation.apply(grid_coords))
```

File: src/morphosamplers/sampler.py [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+    # shift each rotated
+    rotated_shifted = np.stack(rotated, axis=1) + positions
+    return rotated_shifted.reshape(-1, *grid_shape)
+
+
+def sample_volume_with_coordinates(
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:107]:
```suggestion
def sample_volume_at_coordinates(
```
this makes more sense to me

**Review by alisterburt — COMMENTED** [review]
Looking good! I have some concerns about the grid generation but otherwise this is looking good

consider separating grid generation from sampling code, they seem quite separate but can easily come later

File: src/morphosamplers/utils.py (hunk 14 above)
Comment (brisvag) [reply src/morphosamplers/utils.py]: I wrapped all this logic in a separate function cause some things were not working and hard to parse and debug.

File: src/morphosamplers/sampler.py (hunk 15 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:68]: I'm sure there are other ways to frame this, but the way I did it is this:
- `grid_shape`: number of points in the grid along each dimension. So a grid shape of `(5, 5)` spits out a 5 by 5 grid of points
- `grid_spacing`: defines the euclidean distance of the above points.

So if you want a 3D grid to sample subvolumes and you want those subvolumes to be `64x64` pixels with a spacing of `1 * image_pixel_spacing`, you do:

```py
def generate_2D_grid(grid_shape=(64, 64), grid_spacing=(1, 1))
```

Comment (brisvag) [reply src/morphosamplers/sampler.py:68]: Another way to do this would be to provide the euclidean size of the grid and the density, which would then become this for the above:
```py
def generate_2D_grid(grid_size=(32, 32), grid_density=(64, 64))
```

I find the first option more intuitive...

File: src/morphosamplers/sampler.py (hunk 17 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:40]: This is where it shines :P but transposing is really too much xD

**Comment by brisvag** [comment]
Updated based on @alisterburt's comments and added tests for everything!

File: src/morphosamplers/sampler.py (hunk 15 above)
Comment (alisterburt) [reply src/morphosamplers/sampler.py:68]: Ah okay I'm with you! I misunderstood what was going because of the word shape

maybe rename `grid_spacing` to `grid_point_spacing` to differentiate the two? Or I just accept that I'm an idiot :)

File: src/morphosamplers/utils.py (hunk 11 above)
Comment (alisterburt) [reply src/morphosamplers/utils.py]: agree monotonically isn't quite right here for the z vectors, it is right for n but seems unnecessary

**Review by alisterburt — COMMENTED** [review]
last tiny thing then good to go from me! great work here

File: src/morphosamplers/sampler.py (hunk 15 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:68]: Nah it's just that both versions are intuitive depending on how you're approaching :P I don't think adding `point` helps much here to be honest...

**Comment by brisvag** [comment]
One question: as I'm working on the surface spline sampling shenanigans, I also needed to generate "`1D`" grids. In the end, they all share the same mechanisms, so I was thinking of putting it all together like this:

```py
def generate_3D_grid(
    grid_shape: Tuple[int, int, int] = (10, 10, 10),
    grid_spacing: Tuple[float, float, float] = (1, 1, 1),
    squeeze: bool = True,
) -> np.ndarray:
    """
    Generate a 3D sampling grid with specified shape and spacing.

    The grid generated is centered on the origin, has shape (*grid_shape, 3)
    and spacing grid_spacing between neighboring points.

    Parameters
    ----------
    grid_shape : Tuple[int, int, int]
        Shape of the 3D grid.
    grid_spacing : Tuple[float, float, float]
        Spacing between points in the sampling grid.
    squeeze: bool
        Discard degenerate dimensions in the output.

    Returns
    -------
    np.ndarray
        Coordinate of points forming the 2D grid.
    """
    # create indices for x and y with correct spacing
    w, h, d = grid_shape
    dw, dh, dd = grid_spacing
    if w != 1:
        width = np.linspace(-1, 1, w) * w * dw
    else:
        w = 0
    if h != 1:
        height = np.linspace(-1, 1, h) * h * dh
    else:
        h = 0
    if d != 1:
        depth = np.linspace(-1, 1, d) * d * dd
    else:
        d = 0
    # convert to stack form (x, y, z, 3)
    grid = np.stack(np.meshgrid(width, height, depth, indexing="ij"), axis=3)
    return grid
```

Which then would mean that the current `2d` version is done with:
```py
generate_3D_grid(grid_shape=(*grid_shape, 1), grid_spacing=(*grid_spacing, 1))
```

and if I want a `1D` version which goes along Z, i can do:
```py
    grid = generate_3D_grid(grid_shape=(0, 0, sampling_thickness), grid_spacing=(0, 0, sampling_spacing))
```

If one wants to keep degenerate dimensions (for example for sampling subvolumes with a single pixel in one dimensions but without them being merged together) they can pass `squeeze=False`, but I don't expect that to be very used.

Does it make sense or is it a bit unwieldy?

**Comment by alisterburt** [comment]
makes sense to me! I don't mind either way but don't mind the code duplication particularly - I find it easier if I don't have to flit around so much when the logic isn't crazy complicated, would probably keep them separate but up to you

File: src/morphosamplers/sampler.py [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+from .spline import Spline3D
+
+
+def generate_2D_grid(
+    grid_shape: Tuple[int, int] = (10, 10), grid_spacing: Tuple[int, int] = (1, 1)
+) -> np.ndarray:
+    """
+    Generate a 2D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, lying on the xy plane,
```
Comment (kevinyamauchi) [inline src/morphosamplers/sampler.py:19]: I think we should specify the normal vector (`[1, 0, 0]`) as xy plane will mean different things to different people.

File: src/morphosamplers/sampler.py [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def generate_3D_grid(
+    grid_shape: Tuple[int, int, int] = (10, 10, 10),
+    grid_spacing: Tuple[int, int, int] = (1, 1, 1),
+) -> np.ndarray:
+    """
+    Generate a 3D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, has shape (*grid_shape, 3)
```
Comment (kevinyamauchi) [inline src/morphosamplers/sampler.py:50]: I personally don't find `(*grid_shape, 3)` clear notation. Perhaps something like shape `(m * n * p, 3)` for gride_shape `(m, n, p)` (assuming that's correct)?

(same for the 2D grid function)

File: src/morphosamplers/sampler.py (hunk 15 above)
Comment (kevinyamauchi) [reply src/morphosamplers/sampler.py:68]: I am fine with `grid_spacing` since it is described in the docstring.

File: tests/test_sampling.py [hunk 23]
```diff
@@ -0,0 +1,73 @@
+import numpy as np
+from scipy.spatial.transform import Rotation
+
+from morphosamplers.sampler import generate_2D_grid, generate_3D_grid, generate_sampling_coordinates, sample_volume_at_coordinates, sample_volume_along_spline, sample_volume_subvolumes
+from morphosamplers.spline import Spline3D
+
+
+def test_generate_2D_grid():
+    shape = (2, 5)
```
Comment (kevinyamauchi) [inline tests/test_sampling.py:9]: Nit: it would be nice to test the spacing (same for 3d grid)

File: src/morphosamplers/sampler.py [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+    and spacing grid_spacing between neighboring points.
+
+    Parameters
+    ----------
+    grid_shape : Tuple[int, int, int]
+        Shape of the 3D grid.
```
Comment (kevinyamauchi) [inline src/morphosamplers/sampler.py:56]:
```suggestion
        The number of grid points along each axis.
```

File: src/morphosamplers/sampler.py [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+    """
+    Generate a 2D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, lying on the xy plane,
+    has shape (*grid_shape, 3) and spacing grid_spacing between neighboring points.
+
+    Parameters
+    ----------
+    grid_shape : Tuple[int, int]
+        Shape of the 2D grid.
```
Comment (kevinyamauchi) [inline src/morphosamplers/sampler.py:25]:
```suggestion
        The number of grid points along each axis.
```

**Review by kevinyamauchi — COMMENTED** [review]
Thanks for all of the updates, @brisvag ! I think this is pretty much good to go once we settle your question regarding the grid API and the tests are passing.

> One question: as I'm working on the surface spline sampling shenanigans, I also needed to generate "1D" grids. In the end, they all share the same mechanisms, so I was thinking of putting it all together like this:
> ...
> Does it make sense or is it a bit unwieldy?

I am okay with this if you keep convenience functions `generate_2d_grid()` and `generate_3d_grid()` (they can call the the `generate_3d_grid()` function in the background). My concern is that I don't think that is going to be a very intuitive API for most people, so I think we should keep the simple convenience functions.

File: tests/test_sampling.py (hunk 23 above)
Comment (kevinyamauchi) [reply tests/test_sampling.py:9]: Sorry to be pedantic, but as I read it, these tests to check that the spacing between each point is correct (e.g., unequal spacing, but correct end points would pass). That being said, I as mentioned above, this is a nitpick, so I'm fine leaving it as-is if you prefer.

Comment (brisvag) [reply tests/test_sampling.py:9]: Ah I see what you mean. Will update!

File: src/morphosamplers/sampler.py (hunk 15 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:68]: Actually, you were right that this was wrong. it should be `-0.5` to `0.5`... As it was currently, with a shape of `(2, 2)` and spacing of `(1, 1)`, we ended up with a the right shape, but the points had a distance of 4 rather than 2.

Comment (brisvag) [reply src/morphosamplers/sampler.py:68]: Fixed the generation code... it's a bit trickier than I made it at first :P there were 2 main issues:
1. `-1:1` meant we had doubled distances
2. we have to multiply by `shape[axis] - 1`, not just `shape[axis]`, because we have both the edge points

**Comment by brisvag** [comment]
@alisterburt @kevinyamauchi I think this is ready to go in!

File: src/morphosamplers/sampler.py [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+from typing import Tuple
+
+import numpy as np
+from scipy.ndimage import map_coordinates
+from scipy.spatial.transform import Rotation
+
+from .spline import Spline3D
+
+
+def generate_3D_grid(
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:12]:
```suggestion
def generate_3d_grid(
```

File: src/morphosamplers/sampler.py [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+    grid_shape: Tuple[int, int, int] = (10, 10, 10),
+    grid_spacing: Tuple[float, float, float] = (1, 1, 1),
+    squeeze: bool = True,
+) -> np.ndarray:
+    """
+    Generate a 3D sampling grid with specified shape and spacing.
+
+    The grid generated is centered on the origin, has shape (m, n, p, 3) for
+    grid_shape (m, n, p), and spacing grid_spacing between neighboring points.
```
[the remark this answers is not in the document]
Comment (alisterburt) [reply src/morphosamplers/sampler.py:21]: feel free to include or not, feels more obvious to me this way

File: src/morphosamplers/sampler.py [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+        Spacing between points in the sampling grid.
+    squeeze: bool
+        Discard degenerate dimensions in the output.
+
+    Returns
+    -------
+    np.ndarray
+        Coordinate of points forming the 3D grid.
+    """
+    # create indices for x and y with correct spacing
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:37]: outdated? should it be xyz?

File: src/morphosamplers/sampler.py [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+    for dim in range(3):
+        ln = grid_shape[dim]
+        spacing = grid_spacing[dim]
+        if ln != 1:
+            coord = np.linspace(-0.5, 0.5, ln) * (ln - 1) * spacing
+        else:
+            coord = 0
+        coords.append(coord)
+    # convert to stack form (x, y, z, 3)
+    grid = np.stack(np.meshgrid(*coords, indexing="ij"), axis=3)
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:48]: This could be replaced with below which feels a little clearer - I like to keep the names I use for image dimensions separate from the names I use for coordinates which index into those dimensions

```suggestion
    grid = np.indices(grid_shape)
    grid = einops.rearrange(grid, 'zyx d h w -> d h w zyx')
    grid -= np.array(grid_shape) / 2
    grid *= spacing
```

```

File: src/morphosamplers/sampler.py [hunk 30]
```diff
@@ -0,0 +1,223 @@
...
+        The number of grid points along each axis.
+    grid_spacing : Tuple[float, float]
+        Spacing between points in the sampling grid.
+
+    Returns
+    -------
+    np.ndarray
+        Coordinate of points forming the 2D grid.
+    """
+    return generate_3D_grid(grid_shape=(*grid_shape, 1), grid_spacing=(*grid_spacing, 1))
```
[the remark this answers is not in the document]
Comment (alisterburt) [reply src/morphosamplers/sampler.py:76]: could also squeeze with einops ;)

File: src/morphosamplers/sampler.py [hunk 31]
```diff
@@ -0,0 +1,223 @@
...
+    -------
+    np.ndarray
+        Coordinate of points combining all the transformed grids.
+    """
+    rotated = []
+    grid_shape = sampling_grid.shape
+    grid_coords = sampling_grid.reshape(-1, 3)
+    # apply each orientation to the grid and store the result
+    for orientation in orientations:
+        rotated.append(orientation.apply(grid_coords))
```
Comment (alisterburt) [inline src/morphosamplers/sampler.py:106]: using a `Rotation` object here means we should be careful to be very explicit about the expected order of the coordinates in the sampling grid

File: src/morphosamplers/sampler.py [hunk 32]
```diff
@@ -0,0 +1,223 @@
...
+        Array of shape (*grid_shape)
+    """
+    batch, *grid_shape, _ = coordinates.shape
+    # map_coordinates wants transposed coordinate array
+    sampled_volume = map_coordinates(
+        volume, coordinates.reshape(-1, 3).T, order=interpolation_order
+    )
+    # reshape back (need to invert due to previous transposition)
+    # and retranspose to get batch back to the 0th dimension
+    return np.swapaxes(sampled_volume.reshape(*grid_shape, batch), -1, 0)
```
[the remark this answers is not in the document]
Comment (alisterburt) [reply src/morphosamplers/sampler.py:144]: suggestion, not mandatory for sure

File: src/morphosamplers/spline.py [hunk 33]
```diff
@@ -190,3 +196,62 @@ def _get_equidistance_spline_samples(
             raise ValueError("derivative order must be [0, spline_order]")
         u = self._get_equidistance_u(separation)
         return self.sample_spline(u, derivative_order=derivative_order)
+
+
+class Spline3D(NDimensionalSpline):
+    """3D spline model with interpolation of orientations along the spline.
```
Comment (alisterburt) [inline src/morphosamplers/spline.py:202]:
```suggestion
    """3D spline model with a consistent local coordinate system.
```

**Review by alisterburt — COMMENTED** [review]
Looking really really great - suggestions are minor in most cases feel free to either address or not

The only one I'm a bit concerned about is the interaction between `Rotation` objects and the coordinate grids being a little fuzzy, need to be very clear about expected ordering of coordinates in the grids

File: src/morphosamplers/sampler.py (hunk 26 above)
Comment (alisterburt) [reply src/morphosamplers/sampler.py:12]: feels more conventional

**Review by kevinyamauchi — APPROVED** [review]
Thanks, @brisvag ! This looks good to me. Thank you for the additional tests. I think this is good to go once you decide which of @alisterburt 's suggestions to incorporate.

File: src/morphosamplers/sampler.py (hunk 29 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:48]: Awesome! But it's off by 0.5, will fix.

File: src/morphosamplers/sampler.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (alisterburt) [reply src/morphosamplers/sampler.py:21]: don't think so, h and w for height and width of 2d images, third dimension is called depth and is added in front

File: src/morphosamplers/sampler.py (hunk 31 above)
Comment (brisvag) [reply src/morphosamplers/sampler.py:98]: I guess you're referring to the `dwh` issue above, or am I misunderstanding?

**Comment by brisvag** [comment]
I either adrressed or responded to the above comments, should be basically done!

File: src/morphosamplers/sampler.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (alisterburt) [reply src/morphosamplers/sampler.py]: `'xyz ... -> ... xyz'` would move the first axis to the end

[the remark this answers is not in the document]
Comment (brisvag) [reply src/morphosamplers/sampler.py]: sweet!

**Comment by alisterburt** [comment]
Looks like some minor import error here then we're good to go, `einops` -> `einops.rearrange` should sort it!

**Comment by brisvag** [comment]
oopsie!

**Comment by kevinyamauchi** [comment]
Test are now passing, so I will merge. Thanks again, @brisvag !
